Modifies analysis for state

qaphla opened the pull request:
This builds on #84 , using the trace reasoning functions to build a modifies analysis for state changes.

A given trace contains some StateSet entries, and can only write to the addresses (principal, state ID pairs) referenced in these entries. We can therefore find the set of possible addresses touched by a given trace, and extend this also to traceful functions. Most of the base traceful functions that we work with are guaranteed to never modify state, regardless of their input trace, and so we can fairly easily build up the (generally small) set of possibly modified addresses for a given traceful function.

The key lemma that makes use of this allows us to say that if a given traceful function `f` does not modify some address `(prin, sid)`, then the result of `get_state prin sid` will be the same before and after calling `f`. This should allow us to reason about connections between states within the same protocol function, although not yet across different protocol functions.

This is still a draft PR, because while it is fairly complete, it is still messy and lacking documentation (but is ready for review of the overall idea). Some SMTPats could likely also be cleaned up (or added --- I haven't been able to think of a good one for any of the forms of the core lemma, and it would be nice to not have to explicitly call this (`traceful_unmodified_same_state`). The lemma itself also feels a bit awkward in its phrasing, but seems to work well enough for now.

qaphla commented:
Noted by @fabian-hk : The two tmp functions at the end of the file are there just for demonstration purposes, and I plan to remove them before this is finalized --- they can safely be ignored when reviewing, though they might give somewhat of an idea of what does/doesn't work with the lemmas.

TWal reviewed (commented):
I did a quick read, it looks overall good. Some random remarks.

I expect that in the future we will have lemmas for each traceful function stating something like ``forall tr. traceful_modifies f tr `subset` ...`` where `...` is a well-chosen set (e.g. it is a singleton for `set_state`)

It seems that this is not using trace splitting? It is just using trace subtraction?

qaphla commented on src/core/DY.Core.Trace.Modifies.fst:
```diff
@@ -0,0 +1,280 @@
+/// the result of looking up the state at the start and end of that trace (traceful) is the same.
+
+
+let rec trace_modifies_state_was_set (#label_t:Type) (prin:principal) (sid:state_id) (tr:trace_ label_t)
+  : Lemma
+    (ensures mem (prin, sid) (trace_modifies tr) <==> (exists b. state_was_set tr prin sid b))
```
Yes, I also realized this this morning --- probably this concise version is better to use overall

qaphla commented:
> I expect that in the future we will have lemmas for each traceful function stating something like `` forall tr. traceful_modifies f tr `subset` ... `` where `...` is a well-chosen set (e.g. it is a singleton for `set_state`)

Yes, that's about how I envision this being used. For the core functions, I think it is probably best to give exactly what the modifies set is, rather than using subset (especially since for all but `set_state` it is empty), but certainly the lemmas currently in this PR could be converted to a quantified form (I think there could even be something akin to `move_requires` to translate between lemmas with arguments and lemmas where those arguments are moved into a quantified variable in the ensures clause).

> It seems that this is not using trace splitting? It is just using trace subtraction?

Indeed, trace subtraction ended up being sufficient for implementing this feature.

Commit:
Cleaned up requires for trace_find functions

Commit:
Added later lemmas for trace find/search first --- Note that no equivalent holds for find/search last, so they are omitted.

Commit:
Implemented find_event_triggered_at_timestamp with trace_find.

Commit:
Revised get_state_aux to use trace_search_last

Commit:
Modified proof options to deflake

Commit:
Trace arithmetic interaction with trace searching

Commit:
Added initial form of modifies analysis

fabian-hk committed:
feat: add key to crypto predicates along the usages (#93)

* Added key to crypto predicates along the usages

* Proof optimization

* Improve proofs and code cleanup

* Make key in crypto predicates not dependent on the key usage

* Simplified Sign bytes_invariant

TWal committed:
chore: update F* and fix proofs (#98)

mtzguido committed:
chore: update for new F* build (#99)

* nix: update for new F*

* Update F* flake (point to new build)

* Makefiles: no FSTAR_HOME

* update F*

---------

Co-authored-by: Théophile Wallez <theophile@wallez.net>

fabian-hk committed:
feat: extend trace printing with state tags (#100)

Commit:
Cleanup of trace modifies analysis (Still has examples)

Commit:
Unfold is_empty

Commit:
Move some core functions to DY.Core.Trace.Manipulation; clean up unnecessary parts of modifies file

qaphla commented:
At this point, I think the analysis is reasonably working --- I would like ideally to have even better automation, but it does seem that making an assertion about the modifies set of a function appropriately triggers lemmas to break it down into smaller pieces. With more complex functions, I'm sure this doesn't always work fully automatically, but it's at least a good start. There are a few TODO items still --- most prominently:

How should we handle testing? This is sort of an overarching question --- we have our examples directory, but it feels like some of these "trivial" examples are less examples, and more just tests to ensure that we don't break some feature inadvertently. Should they still go with the other examples? Should we split examples into small examples really focused on testing a specific feature, and then larger examples that are more of a "real", if simple, use case? Something else altogether?

In any case, I think the core analysis here is ready to review (i.e., I think it's functionally done, and just needs cleanup), but would also be glad to address feedback that suggests that it isn't. I did also play around with forall-versions of the lemmas for traceful modifies ( I think this is the only fundamental kind of decision that came up from earlier review), but found that they were less performant, so I've dropped them for now.

TWal reviewed (commented):
Looks overall good, some more comments:
- There is a lemma about `traceful_modifies` and `(let*)`, but not with `(let*?)`? I suspect in this version we can't prove of `traceful_modifies` is equal to something, but we can prove it is a subset of something
- Looking at the example, it looks like the SMT pattern of `trace_grows_same_state` is not working terribly well. But I guess it's fine to call a lemma in that situation though
- Does the example still work well if we add an entry (e.g. trigger an event) before the first `get_state`? Or after the last `get_state`?

About tests, indeed the examples serve also serve as tests for DY*, but I guess they are also useful to show how we specify and prove protocols in DY*. For things that look more like unit tests, I think it would be best to create a `test` directory, dedicated for them.

qaphla committed:
Update src/core/DY.Core.Trace.Modifies.fst

Small fix to replace let* with ;* in an example

Co-authored-by: Théophile Wallez <theophile@wallez.net>

qaphla commented on src/core/DY.Core.Trace.Modifies.fst:
```diff
@@ -0,0 +1,297 @@
+    let* st_opt2 = get_state prin sid in
+    return (st_opt1, st_opt2)
+
+#push-options "--z3rlimit 100"
```
Quite possible --- I'll check that  (and also make the example a bit more extensive --- in the end, of course, we want this to work robustly even with non-empty modifies sets). I also didn't change the rlimit in a very fine-grained way, rather just pushed it high enough to make sure it finishes, so once I make the global fuel/ifuel limit for the file, I'll also do some searching to see how low of an rlimit works for the example(s).

qaphla replied:
Indeed, lowering fuel/ifuel drastically speeds the example up.

qaphla commented on src/core/DY.Core.Trace.Modifies.fst:
```diff
@@ -0,0 +1,297 @@
+    ))
+    [SMTPat (get_state prin sid tr1);
+     SMTPat (get_state prin sid tr2);
+     SMTPat (~((prin, sid) `mem` (trace_modifies (tr2 <--> tr1))));
```
I didn't get a warning about it when running locally (without the build process), but maybe I have different flags set or something. I guess the solution here would be to define a new name for this not-membership predicate, and use that named predicate in the SMTPat? I had also thought about doing this regardless, as I think it makes the preconditions of the lemmas look nicer.

I also think in this particular case, the SMTPat is perhaps not so useful --- it didn't work as well as I had hoped for examples, but then maybe that's partly because of the `~`. After that change, if it still isn't working smoothly, I would probably just remove the pattern.

qaphla replied:
This is now improved somewhat, and at least does appear to get triggered in the proof of `traceful_unmodified_same_state`. Ideally, though, I'd really like to find a working pattern for the latter lemma, since that's the one I expect to be most often relevant.

qaphla commented:
>     * There is a lemma about `traceful_modifies` and `(let*)`, but not with `(let*?)`? I suspect in this version we can't prove of `traceful_modifies` is equal to something, but we can prove it is a subset of something

Probably a `(let*?)` looking similar to the `(let*)` one is possible. I'll add this and see how it works. Indeed, I expect in this version it would only be able to prove a subset, which I think should be sufficient for what we actually need in the end, but then the question is how well that will interact with the `let*` lemma's use of equality


>     * Does the example still work well if we add an entry (e.g. trigger an event) before the first `get_state`? Or after the last `get_state`?

I suspect the proof needs to be made more complex in that case, but I can test it.
 
> About tests, indeed the examples serve also serve as tests for DY*, but I guess they are also useful to show how we specify and prove protocols in DY*. For things that look more like unit tests, I think it would be best to create a `test` directory, dedicated for them.

I'll pull the test case out into such a directory then. `test/core/DY.Core.Trace.Modifies.Test.fst`, probably, mirroring the structure of the `src` directory.

Commit:
Added fuel restriction to modifies file, added modifies to core

Commit:
Added predicates for addresses being unmodified by a trace/traceful; added traceful option bind modifies lemma

Commit:
Moved initial test for modifies analysis into a testing directory. To be further moved if needed.

Commit:
Fixed up testing for modifies analysis, with a couple basic tests.

TWal committed:
feat: add support for MAC (#103)

* feat: add support for MAC

* sprinkle a few comments

* `mac_auth` -> `mac_compute`

qaphla committed:
cleanup: trace find/search (#96)

* Cleaned up requires for trace_find functions

* Added later lemmas for trace find/search first --- Note that no equivalent holds for find/search last, so they are omitted.

* Implemented find_event_triggered_at_timestamp with trace_find.

* Revised get_state_aux to use trace_search_last

* Modified proof options to deflake

* Trace arithmetic interaction with trace searching

* Uniformize SMTPats on growing traces

For various `later` lemmas in Trace.Base, the SMT pattern has been weakened for consistency with other such lemmas.

---------

Co-authored-by: Klaas Pruiksma <klaas.pruiksma@sec.uni-stuttgart.de>

TWal committed:
feat: add SMT patterns for Comparse, with a switch (#105)

TWal committed:
cleanup: reduce boilerplate to merge local predicates (#104)

fabian-hk committed:
feat: add Communication Layer (#50)

Co-authored-by: Klaas Pruiksma <qaphla@users.noreply.github.com>
Co-authored-by: Klaas Pruiksma <klaas.pruiksma@sec.uni-stuttgart.de>

qaphla committed:
Merge branch 'main' into qaphla/modifies-analysis

Commit:
Fix up flakiness

Commit:
Update makefile to build test directories

We may want to further update to treat tests separately --- e.g., not building them in the main build command.

TWal commented on src/core/DY.Core.Trace.Manipulation.fst:
```diff
@@ -547,7 +571,7 @@ val trigger_event: principal -> string -> bytes -> traceful unit
   add_entry (Event prin tag content)
 
-#push-options "--z3rlimit 30"
+#push-options "--z3rlimit 25 --ifuel 2"
```
I am curious why this needs an increase in ifuel?

TWal replied:
Ah interesting, I just hit the same flakiness, and although increasing the rlimit works in the interactive mode, it does not work when running `make`, however increasing ifuel works... weird, probably an F* bug

TWal replied:
In one of the recent merged PRs I put a better fix (this theorem proves an `exists`, so I give a witness rather than letting Z3 guessing it)

TWal commented on src/core/DY.Core.Trace.Modifies.fst:
```diff
@@ -0,0 +1,346 @@
+  Lemma
+  (requires tr1 <$ tr2 /\ trace_does_not_modify_addr prin sid (tr2 <--> tr1))
+  (ensures (get_state_aux prin sid tr1) == (get_state_aux prin sid tr2))
+  [SMTPat (tr1 <$ tr2);
+   SMTPat (get_state_aux prin sid tr2);
+   SMTPat (get_state_aux prin sid tr1);
+  ]
```
I wonder about the lemmas with SMT patterns on `get_state_aux`, is this supposed to be used by the user, or is just internal to the library?
If it is supposed to be used by the user, does it mean they must reveal the definition of `get_state`?

qaphla replied:
This approach seems to work quite well --- I'll push an update soon with that version. One question that arises: What should we call this predicate `p`? I'm currently using `state_most_recent` in my draft, but I don't like that very much. It seems like an ideal name would capture:
1. That it is about state
2. Something about this being the most recent state for a given address
but also make sense in the case where there is no state at that address (i.e., `p prin sid None tr` should sound reasonable). All ideas welcome.

TWal replied:
What about something like `get_state_returns` or `get_state_would_return`?

qaphla replied:
I've made a change that adds this. It seems to mostly work nicely, but the automation fails in essentially the same place as the old version: when trying to propagate information across traceful functions that do not modify a given address. The lemma `traceful_get_state_would_return_later` works well for this (much more nicely than the old version, I think --- I would be inclined to remove the old-style lemmas entirely, if no one objects), but I haven't been able to find a way to reliably trigger that lemma in real proofs. One problem is that writing a lemma over a generic `f : traceful a` seems to go poorly with patterns --- the pattern `f tr` does not bind `a`, it seems, and so F* is unhappy with this as a pattern unless something else does more explicitly bind the type `a`, as the current pattern `traceful_does_not_modify_addr prin sid f tr` does. However, needing this as a pattern is bad for automation --- I don't see a good way to introduce this into the context without asserting it explicitly.

That said, it does seem to work no worse than the old version, and is much more elegant.

TWal replied:
I agree it is better! I will try to think a bit about the SMT pattern issue and see if I come up with anything useful, but it does look quite hairy indeed.

TWal reviewed (dismissed):
Looks overall good!

qaphla commented on test/core/DY.Test.Core.Trace.Modifies.fst:
```diff
@@ -0,0 +1,83 @@
+  = let* new_rand = mk_rand NoUsage (DY.Core.Label.principal_label prin) 32 in
+    let* msg_ts = send_msg new_rand in
+    trigger_event prin "test_event" (Literal (Seq.Base.empty));*
+//  Adding this next line back in causes the automation below to fail.
+//  Something to do with needing to search too deep?
+//    let* _ = recv_msg msg_ts in
```
It does indeed work with that setting, and even can have more extraneous lines added --- this gets it up to 28 lines worth of extra steps before it stops working.

Commit:
Added second branch test, increased qi threshold in testing

Commit:
Add get_state_would_return predicate

This predicate allows us to talk about what get_state(_aux) would yield without needing
to expose the definition of get_state.

Some lemmas are added to prove this predicate after set_state, to propagate it through
the trace (somewhat poorly so far), and to use it to draw conclusions about the result
of get_state.

Commit:
Removes get_state_aux-referencing lemmas from modifies.

Also updates modifies tests to work with get_state_would_return.

qaphla committed:
Merge branch 'main' into qaphla/modifies-analysis

qaphla committed:
Merge branch 'main' into qaphla/modifies-analysis

Commit:
Pull predicates/lemmas  on most recent state from state update branch

Commit:
Update makefile to more carefully include test directories.

Commit:
Updated modifies analysis to use is_state_for and is_most_recent_state_for

qaphla commented:
The most recent sets of changes here are largely cleanup and unifying with #112 --- both depend on a lot of the same core predicates for reasoning, to do with what state was most recently set, and so those predicates and many of their associated lemmas seem to belong in more core places.

@TWal Are there further objections you have/should anything be renamed again? I think the biggest obvious change since you last looked is that `get_state_would_return` is now `is_most_recent_state_for`, which I still don't find ideal, but while `get_state_would_return` makes sense in the context of the modifies work, it feels a very unnatural name for #112, where it's used in the trace invariant, before `get_state` is even defined. There is also a ghost version of `get_state_aux` used in defining that predicate, which might be another thing to look at.

Automation is still (as before) not quite to the point where I want it, but if other objections are all addressed, I'd like to have this merged so that #112 can more easily build on it without it being too fragile a tower of PRs, and automation improvements can both come later and be a smaller change.

Opinions from anyone else also of course welcome.

Commit:
Removed some no longer needed functions about get_state_aux

qaphla commented:
This most recent commit cleans up some functions that were needed in an older version and didn't get properly removed in the newest version.

qaphla committed:
Merge branch 'main' into qaphla/modifies-analysis

Commit:
Update tagged state with modifies

Commit:
Added modifies lemma for typed events.

Commit:
Added modifies code to typed state.

Commit:
Merge branch 'qaphla/modifies-analysis' of github.com:REPROSEC/dolev-yao-star-extrinsic into qaphla/modifies-analysis

qaphla committed:
Merge branch 'main' into qaphla/modifies-analysis

qaphla committed:
Merge branch 'main' into qaphla/modifies-analysis

Commit:
Deflake

Commit:
Cleaned up TODOs in modifies.

qaphla commented:
This should now have all TODOs removed and be ready to review @TWal .

Automation is still not as good as would be ideally desired, but it is at least workable.
Adding `traceful a` to the pattern for `traceful_is_most_recent_state_for_later` didn't seem to make it trigger any better. I'm not entirely sure how to debug this, since even if I put all three of the patterns into the context in the test file, it just doesn't activate. Is there a flag to pass to see what's going on with patterns? I suppose one other possibility is that adding that pattern caused it to be triggered too much, but not for the right purposes (i.e., trying to propagate the information about every address everywhere, not only for the relevant addresses and to where they need to go). Still, that can be resolved in a future PR.

TWal reviewed (commented):
Looks mergeable, I only have a few nits on the test file. But otherwise, excellent work, thanks!

Commit:
Cleaned up modifies tests, making recv_msg optional

qaphla commented on test/core/DY.Test.Core.Trace.Modifies.fst:
```diff
@@ -0,0 +1,151 @@
+
+/// Without this option, the proof automation gets stuck working through
+/// "extraneous" steps.
+#push-options "--z3cliopt 'smt.qi.eager_threshold=100'"
```
At this point, I forget what the reason for that comment was, so indeed probably best to remove it, especially as it's not very clear.

qaphla commented:
Nits on the test file should be resolved now --- all instances of `recv_msg` were made optional, the qi threshold is set for the whole file, and the related comment is removed.
